pennylane/devices/qubit/preprocess.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains functions for preprocessing `QuantumScript`s to ensure
+that they are supported for execution by a device."""
+
+import pennylane as qml
+
+# from pennylane.tape import QuantumScript
+from pennylane.operation import Observable, Tensor
+from pennylane.measurements import (
+    MidMeasureMP,
+    ExpectationMP,
+    ShadowExpvalMP,
+)
+from pennylane import DeviceError
+
+
+def _stopping_condition(op):
+    """Specify whether or not an Operator object is supported by the device"""
+    return getattr(op, "has_matrix", False) and qml.matrix(op).shape[-1] % 2 == 0
```

Comment:
> Is `qml.matrix(op).shape[-1] % 2 == 0` a proxy for figuring out if its qubit or qutrit?

Precisely. I think adding a `level` class attribute to the devices and operators would be a good short-term (but potentially also long-term) solution such that the matrix doesn't need to be evaluated to differentiate between qubits and qutrits. That would also allow better generalization for many operations that are (or would make sense to be) shared between qubit and qutrit devices, such as `SWAP`, `Identity`, or higher dimensional generalizations of the `Pauli` and `Hadamard` operators.

For now, this check is important because otherwise `preprocess` would validate qubit circuits with qutrit operations.

Reply:
Let's just avoid this qutrit check for now.  We can add something like this back in when we have a quick way to distinguish qubit from qutrit.